### PR TITLE
🌱 Use ClusterClass name index in ClusterClass webhook

### LINKED ISSUE #5719: Add list by index to ClusterClass webhook
The ClusterClass webhook  in PR #5717 lists all clusters and then cycles through them to find those using the clusterClass being validated. Instead the ClusterClass webhook should use index.ClusterClassNameField to pull the correct clusters directly. 

This workflow is currently blocked as it requires envtest to test the indices. fakeClient does not populate indices  like index.ClusterClassNameField. The index currently can't be tested as there is a circular dependency between webhooks and envtest.

The index should be added once the circular dependency with webhooks has been removed.

c/f https://github.com/kubernetes-sigs/cluster-api/pull/5717#discussion_r754532743

### PR DESCRIPTION
**Pull request by killianmuldoon**
Signed-off-by: killianmuldoon <kmuldoon@vmware.com>

Add the ClusterClass name index to the client call for getting a ClusterClass in the ClusterClass webhook. 

Certain tests are moved to a new subpackage in order to run them with envtest. This is required as the fakeclient doesn't set up the indexes now in use by the ClusterClass webhook.

Note this PR includes duplicate files from #5803 and should not be merged until that PR is completed.

Fixes #5719

### CODE REVIEW

File: internal/webhooks/test/clusterclass_test.go [hunk 1]
```diff
@@ -28,234 +28,250 @@ import (
 	"sigs.k8s.io/cluster-api/internal/test/builder"
 )
```
Comment (sbueringer) [inline internal/webhooks/test/clusterclass_test.go:30]: Q: Can you give some context on why the old test cases where removed? Are they included in the new ones?

File: internal/webhooks/test/clusterclass_test.go [hunk 2]
```diff
@@ -28,234 +28,250 @@ import (
...
-// In this case creation of the ClusterClass should be blocked by the webhook.
-func TestClusterClassWebhook_Fail_Create(t *testing.T) {
-	defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.ClusterTopology, true)()
+// TestClusterClassWebhook_Succeed_Delete_ClusterClassWithSimilarMachineDeploymentClass tests the correct deletion behaviour for a ClusterClass with no references in existing Clusters.
+// In this case deletion of the ClusterClass should succeed.
+func TestClusterClassWebhook_Succeed_Delete_ClusterClassWithSimilarMachineDeploymentClass(t *testing.T) {
 	g := NewWithT(t)
 
-	ns, err := env.CreateNamespace(ctx, "test-topology-clusterclass-webhook")
+	ns, err := env.CreateNamespace(ctx, "test-topology-cluster-reconcile")
```
Comment (sbueringer) [inline internal/webhooks/test/clusterclass_test.go:36]: Not sure if the new namespace name is correct

File: internal/webhooks/test/clusterclass_test.go [hunk 3]
```diff
@@ -28,234 +28,250 @@ import (
...
-						WithClass("IN_USE").
-						Build(),
-				).
-				Build()).
-		Build()
-
-	// Create the ClusterClass in the API server.
-	g.Expect(env.CreateAndWait(ctx, clusterClass)).To(Succeed())
+	// - two ClusterClasses with all the related templates and very similar MachineDeployments
+	// - a Cluster using one of the ClusterClasses
```
Comment (sbueringer) [inline internal/webhooks/test/clusterclass_test.go:43]: looks like two clusters

File: internal/webhooks/test/clusterclass_test.go [hunk 4]
```diff
@@ -28,234 +28,250 @@ import (
...
-
-// TestClusterWebhook_Succeed_Update tests the correct update behaviour for a ClusterClass with references in existing Clusters.
-// In this case deletion of the ClusterClass should succeed by the webhook as the update is compatible with the existing Clusters using the ClusterClass.
-func TestClusterWebhook_Succeed_Update(t *testing.T) {
-	defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.ClusterTopology, true)()
-	g := NewWithT(t)
-
-	ns, err := env.CreateNamespace(ctx, "test-topology-clusterclass-webhook")
-	g.Expect(err).ToNot(HaveOccurred())
+	similarMachineDeploymentClassName := "aa"
```
Comment (sbueringer) [inline internal/webhooks/test/clusterclass_test.go:39]: This var is pretty confusing to me.

So my summary of this test would be:
* we create two ClusterClasses:
   * class1: with MD classes: aa, bb
   * class2: with MD classes: aa, bb
* we create two Clusters:
   * cluster1 uses class1 and MD class bb
   * cluster2 uses class2 and MD class aa

Then we delete MD class aa in ClusterClass class1 which is fine as we don't have a cluser using that MD class. At least for me, it feels like it's to much effort to figure that out based on the code.

File: internal/webhooks/test/clusterclass_test.go (hunk 1 above)
Comment (killianmuldoon) [reply internal/webhooks/test/clusterclass_test.go:31]: Switching to index means that none of the Delete cases works in tests using fakeclient. Indexes only work with testenv.

All of the delete test cases were removed (I moved over the remaining ones in the latest change) and should be covered by tests in the new internal/webhooks/test package.

File: internal/webhooks/test/clusterclass_test.go (hunk 4 above)
Comment (killianmuldoon) [reply internal/webhooks/test/clusterclass_test.go]: I've removed this test in the latest version.

File: internal/webhooks/test/clusterclass_test.go (hunk 1 above)
Comment (sbueringer) [reply internal/webhooks/test/clusterclass_test.go:31]: I meant why the cases in this package (which are using envtest) where dropped. I assume they are covered by the new ones.

Comment (killianmuldoon) [reply internal/webhooks/test/clusterclass_test.go]: Just took a look - those tests got removed when I was doing the merge / rebase. Thanks for catching that. I've added them back in the latest version of the code.

Comment (sbueringer) [reply internal/webhooks/test/clusterclass_test.go]: Ah good. I was trying to figure out how they could be part of the new ones :joy:

**Comment by sbueringer** [comment]
@killianmuldoon thx! lgtm pending squash

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**